docs/lake/design.md

```diff
@@ -0,0 +1,657 @@
+
+```
+lake-<lake-id>/
+  config/
+    HEAD
+    0.zng
+    1.zng
+    2.zng
+    ...
+  index-rules/
+    HEAD
+    0.zng
+    1.zng
+    2.zng
+    ...
+  pool-<pool-id>/
+    config/
+      TAIL
+      0.zng
+      1.zng
+      ...
+    journal/
+      HEAD
+      TAIL
+      0.zng
+      1.zng
+      2.zng
+      ...
+      20.zng
+      20-snap.zng
+      20-seek.zng
+      21.zng
+      ...
+    data/
+      <tag>-0-.{zng,zst}
+      <tag>-0-seek.zng
+      <tag>-0-index-0.zng
+      <tag>-0-index-1.zng
+      ...
+      <tag>-1-.{zng,zst}
+      <tag>-1-seek.zng
+      <tag>-1-index-0.zng
+      <tag>-1-index-1.zng
+      ...
+  pool-<pool-id>/
+  ...
+```
+
+## Continuous Ingest
+
+While the description above is very batch oriented, the Zed like design is
+intended to perform scalably for continuous streaming applications.  In this
+approach, many small commits may be continuously executed as data arrives and
+after each commit, the data is immediately readable.
+
+To handle this use case, the _journal_ of commits is design
+to scale to arbitrarily large footprints as described earlier.
+
+## Lake Introspection
+
+Commit history, meta data about segments, segment key spaces,
+etc can all be queried and
+returned as Zed data, which in turn, can be fed into Zed analytics.
+This allows a very powerful approach to introspecting the structure of a
+lake making it easy to measure, tune, and adjust lake parameters to
+optimize layout for performance.
+
+> TBD: define model for scanning metadata in this fashion.  It might be as
+> easy as scanning virtual sub-pools that conform to the different types of
+> metadata related to a pool, e.g., logs.$segments, logs.$commits, etc.
+
+## Derived Analytics
+
+To improve the performance of predictable workloads, many use cases of a
+Zed lake pre-compute _derived analytics_ or a particular set of _partial
+aggregations_.
+
+For example, the Brim app displays a histogram of event counts grouped by
+a category over time.  The partial aggregation for such a computation can be
+configured to run automatically and store the result in a pool designed to
+hold such results.  Then, when a scan is run, the Zed analytics engine
+recognizes when the DAG of a query can be rewritten to assemble the
+partial results instead of deriving the answers from scratch.
+
+When and how such partial aggregations are performed is simply a matter of
+writing Zed queries that take the raw data and produce the derived analytics
+while conforming to a naming model that allows the Zed lake to recognize
+the relationship between the raw data and the derived data.
+
+> TBD: Work out these details which are reminiscent of the analytic cache
+> developed in our earlier prototype.  For example,
+
+## Keyless Data
+
+This is TBD.  Data without a key should be accepted some way or another.
+One approach could be to simply assign the "zero-value" as the pool key.
+Or a configured default value could be user.  This would make key-based
```

Comment:
```suggestion
One approach could be to simply assign the "zero-value" as the pool key,
or a configured default value could be used.  This would make key-based
```